packages/@react-aria/grid/src/useGrid.ts

```diff
@@ -98,6 +98,35 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
     scrollRef
   });
 
+  const cachedFocusedKey = useRef(null);
+  const cachedCollection = useRef(null);
+  useEffect(() => {
+    if (
+      state.collection.size > 0 &&
+      state.selectionManager.focusedKey === null &&
+      cachedFocusedKey.current !== null
+    ) {
+      const node = cachedCollection.current.getItem(cachedFocusedKey.current);
+      const isCell = node.type === 'cell' || node.type === 'rowheader' || node.type === 'column';
+      const parentNode =
+        node.parentKey && isCell ?
+        cachedCollection.current.getItem(node.parentKey) :
+        cachedCollection.current.getItem(node.key);
+      const rows = state.collection.rows;
+      const newRow = parentNode.index < rows.length ? rows[parentNode.index] : rows[rows.length - 1];
+      const keyToFocus =
+        newRow && newRow.hasChildNodes && isCell ?
+        [...newRow.childNodes][node.index].key :
+        newRow.key;
+      if (keyToFocus) {
+        state.selectionManager.setFocusedKey(keyToFocus);
+      }
+    }
+    cachedFocusedKey.current = state.selectionManager.focusedKey;
+    cachedCollection.current = state.collection;
+  },
+  [state.collection, state.selectionManager, state.selectionManager.focusedKey]);
```

Comment:
Now that this doesn't depend on the keyboard delegate, maybe we can centralize this logic across components. Both `useGridState` and `useListState` already have some logic that sets the focused key to null if it gets deleted. Those should probably be replaced with this. Moving it to `useListState` would also mean that other components like `ListBox` would also get this behavior.

Reply:
I have refactored the code into useGridState and useListState.